### PR TITLE
Adding support for structure patterns

### PR DESCRIPTION
**Pull request by vilda-r7**

### CODE REVIEW

File: src/le.py [hunk 1]
```diff
@@ -3372,12 +3531,184 @@ def cmd_ls(args):
...
+        print >> sys.stderr, 'Added structure `%s\'' % structure_name
+        structure_id = response['structure']['id']
+    else:
+        structure_id = matches[0]
+        if not pattern:
+            print >> sys.stderr, 'Structure `%s\' already exists' % structure_name
+
+    # Add pattern (if specified)
+    if pattern:
+
```
Comment (fpedrini) [inline src/le.py]: I'd remove the blank lines here, and move the comment above the `if`, it makes it a bit harder to read otherwise.

File: src/le.py [hunk 2]
```diff
@@ -3372,12 +3531,184 @@ def cmd_ls(args):
...
+    rm_all = pattern in ['*', '.']
+
+    # Load all patterns
+    response = api_v2_request('GET', 'structures/%s/patterns'%structure_id, {}, True)
+    patterns = response['patterns']
+
+    # Go though patterns one by one, identify pattern IDs
+    # Use a different call for pattern=*
+    matches = [[p['id'], p['pattern']] for p in patterns if rm_all or p['id'] == pattern or p['pattern'].startswith(pattern)]
```
Comment (fpedrini) [inline src/le.py]: I get the list comprehension but there are three different conditions, which makes it hard to parse at a glance. I'd be in favour of using for and ifs.

File: src/utils.py [hunk 3]
```diff
@@ -17,14 +17,33 @@
 
 import logging
 
+try:
+    import termcolor
```
Comment (fpedrini) [inline src/utils.py]: ❤️

File: src/le.py [hunk 4]
```diff
@@ -3372,12 +3531,184 @@ def cmd_ls(args):
...
+        structure_name = args[1]
+        patterns = args[2:]
+    elif subject.startswith('structures/'):
+        structure_name = subject[len('structures/'):]
+        patterns = args[1:]
+
+    if not structure_name:
+        error('No structure name specified (append structure name)')
+
+    # Not args contain structure name or ID and an optional pattern
```
Comment (fpedrini) [inline src/le.py]: ~ is this a leftover? or it should be `Note: ...`?

File: src/le.py [hunk 5]
```diff
@@ -3347,12 +3421,97 @@ def cmd_ls_ips(ags):
...
+        cmd_ls_ips()
+        return
+    if len(args) == 1 and args[0] == 'structures':
+        cmd_ls_structures()
+        return
+    if len(args) >= 1 and args[0].startswith('structures/'):
+        if len(args) == 1:
+            cmd_ls_patterns(args[0][len('structures/'):])
+        else:
+            die('Error: Too many arguments.')
```
Comment (fpedrini) [inline src/le.py]: In the other `cmd_*` methods in this PR `error()` is called when there are missing/wrong arguments, should this be colored as well?

File: src/le.py [hunk 6]
```diff
@@ -2388,6 +2410,58 @@ def api_request(request, required=False, check_status=False, silent=False, die_o
...
+        reason = ''
+        if 'reason' in d_response:
+            reason = d_response['reason']
+        if required:
+            if reason:
+                die("Error: %s" % reason)
+            else:
+                die("Error: Cannot process LE request: (%s)" % response.status)
+        return None
+
```
Comment (fpedrini) [inline src/le.py]: Too many blank lines here imho.

File: src/le.py [hunk 7]
```diff
@@ -2388,6 +2410,58 @@ def api_request(request, required=False, check_status=False, silent=False, die_o
...
+    """
+    # Obtain response
+    request_encoded = json.dumps(request)
+    response, conn = get_response(
+        method, '/v2/accounts/%s/%s'%(config.user_key, url), request_encoded,
+        silent=silent, die_on_error=die_on_error, domain=Domain.API,
+        headers={'Content-Type': 'application/x-www-form-urlencoded'})
+
+    # Check the response
+    if not response:
```
Comment (vilda-r7) [inline src/le.py]: Using contextlib would be very nice. However, it is in Python from version 2.5 and we have to support Python 2.4 due to the wide spread of systems we support.

File: src/le.py [hunk 8]
```diff
@@ -2388,6 +2410,58 @@ def api_request(request, required=False, check_status=False, silent=False, die_o
...
+            d_response = {}
+    except ValueError:
+        error = 'Error: Invalid response, parse error.'
+        if die_on_error:
+            die(error)
+        else:
+            log.info(error)
+            d_response = None
+
+    if response.status not in [200, 201, 204]:
```
Comment (vilda-r7) [inline src/le.py]: Well, this is a list. Let me know if you want something else. I'll use httplib constants, thanks

File: src/le.py (hunk 6 above)
Comment (vilda-r7) [inline src/le.py]: Removed

File: src/le.py [hunk 9]
```diff
@@ -3347,12 +3421,97 @@ def cmd_ls_ips(ags):
     print ' '.join(l)
 
 
+def cmd_ls_structures():
```
Comment (vilda-r7) [inline src/le.py]: I agree. Though there's push to use the other command line tool and these commands will be removed - we'll see.

File: src/le.py [hunk 10]
```diff
@@ -3347,12 +3421,97 @@ def cmd_ls_ips(ags):
...
+
+    if len(structures) == 0:
+        print >> sys.stderr, 'No structures defined'
+    elif len(structures) == 1:
+        print >> sys.stderr, '1 structure'
+    else:
+        print >> sys.stderr, '%s structures' % len(structures)
+
+
+def cmp_patterns(a, b):
```
Comment (vilda-r7) [inline src/le.py]: Moved now

File: src/le.py (hunk 5 above)
Comment (vilda-r7) [inline src/le.py]: I used error function in new code only to make the PR smaller. I extended its use to other cases now.

File: src/le.py (hunk 1 above)
Comment (vilda-r7) [inline src/le.py]: okay

File: src/le.py (hunk 2 above)
Comment (vilda-r7) [inline src/le.py]: I split it to two lines, it  should be easy to read now

File: src/le.py (hunk 4 above)
Comment (vilda-r7) [inline src/le.py]: It supposed to be Now. I change it to Note and added a comma to separate conditions

**Comment by vilda-r7** [comment]
Thanks for comment @fpedrini all are addressed in latest update

**Comment by vilda-r7** [comment]
Since this is POC that would eventually be moved to official cli, creating a server mock would be quite an overshot I believe

**Comment by udangel-r7** [comment]
@ellynch-r7 can you have a look?
